File: ruby/intp/test/test_lexer.rb
require_relative 'helper'

class LexerTest < Minitest::Test
  def test_next_token
    input = <<INPUT
    let five = 5;
    let ten = 10;

    let add = fn(x, y) {
      x + y;
    };

    let result = add(five, ten);
    !-/*5;
    5 < 10 > 5;

    if (5 < 10) {
      return true;
    } else {
      return false;
    }

    10 == 10;
    10 != 9;
    "foobar"
    "foo bar"
    [1, 2];
    {"foo": "bar"}
INPUT

    tests = [
      [Intp::Token::LET, 'let'],
      [Intp::Token::IDENT, 'five'],
      [Intp::Token::ASSIGN, '='],
      [Intp::Token::INT, '5'],
      [Intp::Token::SEMICOLON, ';'],
      [Intp::Token::LET, 'let'],
      [Intp::Token::IDENT, 'ten'],
      [Intp::Token::ASSIGN, '='],
      [Intp::Token::INT, '10'],
      [Intp::Token::SEMICOLON, ';'],
      [Intp::Token::LET, 'let'],
      [Intp::Token::IDENT, 'add'],
      [Intp::Token::ASSIGN, '='],
      [Intp::Token::FUNCTION, 'fn'],
      [Intp::Token::LPAREN, '('],
      [Intp::Token::IDENT, 'x'],
      [Intp::Token::COMMA, ','],
      [Intp::Token::IDENT, 'y'],
      [Intp::Token::RPAREN, ')'],
      [Intp::Token::LBRACE, '{'],
      [Intp::Token::IDENT, 'x'],
      [Intp::Token::PLUS, '+'],
      [Intp::Token::IDENT, 'y'],
      [Intp::Token::SEMICOLON, ';'],
      [Intp::Token::RBRACE, '}'],
      [Intp::Token::SEMICOLON, ';'],
      [Intp::Token::LET, 'let'],
      [Intp::Token::IDENT, 'result'],
      [Intp::Token::ASSIGN, '='],
      [Intp::Token::IDENT, 'add'],
      [Intp::Token::LPAREN, '('],
      [Intp::Token::IDENT, 'five'],
      [Intp::Token::COMMA, ','],
      [Intp::Token::IDENT, 'ten'],
      [Intp::Token::RPAREN, ')'],
      [Intp::Token::SEMICOLON, ';'],
      [Intp::Token::BANG, '!'],
      [Intp::Token::MINUS, '-'],
      [Intp::Token::SLASH, '/'],
      [Intp::Token::ASTERISK, '*'],
      [Intp::Token::INT, '5'],
      [Intp::Token::SEMICOLON, ';'],
      [Intp::Token::INT, '5'],
      [Intp::Token::LT, '<'],
      [Intp::Token::INT, '10'],
      [Intp::Token::GT, '>'],
      [Intp::Token::INT, '5'],
      [Intp::Token::SEMICOLON, ';'],
      [Intp::Token::IF, 'if'],
      [Intp::Token::LPAREN, '('],
      [Intp::Token::INT, '5'],
      [Intp::Token::LT, '<'],
      [Intp::Token::INT, '10'],
      [Intp::Token::RPAREN, ')'],
      [Intp::Token::LBRACE, '{'],
      [Intp::Token::RETURN, 'return'],
      [Intp::Token::TRUE, 'true'],
      [Intp::Token::SEMICOLON, ';'],
      [Intp::Token::RBRACE, '}'],
      [Intp::Token::ELSE, 'else'],
      [Intp::Token::LBRACE, '{'],
      [Intp::Token::RETURN, 'return'],
      [Intp::Token::FALSE, 'false'],
      [Intp::Token::SEMICOLON, ';'],
      [Intp::Token::RBRACE, '}'],
      [Intp::Token::INT, '10'],
      [Intp::Token::EQ, '=='],
      [Intp::Token::INT, '10'],
      [Intp::Token::SEMICOLON, ';'],
      [Intp::Token::INT, '10'],
      [Intp::Token::NOT_EQ, '!='],
      [Intp::Token::INT, '9'],
      [Intp::Token::SEMICOLON, ';'],
      [Intp::Token::STRING, 'foobar'],
      [Intp::Token::STRING, 'foo bar'],
      [Intp::Token::LBRACKET, '['],
      [Intp::Token::INT, '1'],
      [Intp::Token::COMMA, ','],
      [Intp::Token::INT, '2'],
      [Intp::Token::RBRACKET, ']'],
      [Intp::Token::SEMICOLON, ';'],
      [Intp::Token::LBRACE, '{'],
      [Intp::Token::STRING, 'foo'],
      [Intp::Token::COLON, ':'],
      [Intp::Token::STRING, 'bar'],
      [Intp::Token::RBRACE, '}'],
      [Intp::Token::EOF, '']
    ]

    l = Intp::Lexer.new(input)

    tests.each do |test|
      token = l.next_token
      assert_equal test[0], token.type
      assert_equal test[1], token.literal
    end
  end
end

File: python/tests/test_lexer.py
import unittest
from intp import token
from intp import lexer


class TestNextToken(unittest.TestCase):
    def test_next_token(self):
        input = """
        let five = 5;
        let ten = 10;
        let add = fn(x, y) {
          x + y;
        };

        let result = add(five, ten);
        !-/*5;
        5 < 10 > 5;
        """

        tests = [
            (token.LET, "let"),
            (token.IDENT, "five"),
            (token.ASSIGN, "="),
            (token.INT, "5"),
            (token.SEMICOLON, ";"),
            (token.LET, "let"),
            (token.IDENT, "ten"),
            (token.ASSIGN, "="),
            (token.INT, "10"),
            (token.SEMICOLON, ";"),
            (token.LET, "let"),
            (token.IDENT, "add"),
            (token.ASSIGN, "="),
            (token.FUNCTION, "fn"),
            (token.LPAREN, "("),
            (token.IDENT, "x"),
            (token.COMMA, ","),
            (token.IDENT, "y"),
            (token.RPAREN, ")"),
            (token.LBRACE, "{"),
            (token.IDENT, "x"),
            (token.PLUS, "+"),
            (token.IDENT, "y"),
            (token.SEMICOLON, ";"),
            (token.RBRACE, "}"),
            (token.SEMICOLON, ";"),
            (token.LET, "let"),
            (token.IDENT, "result"),
            (token.ASSIGN, "="),
            (token.IDENT, "add"),
            (token.LPAREN, "("),
            (token.IDENT, "five"),
            (token.COMMA, ","),
            (token.IDENT, "ten"),
            (token.RPAREN, ")"),
            (token.SEMICOLON, ";"),
            (token.BANG, "!"),
            (token.MINUS, "-"),
            (token.SLASH, "/"),
            (token.ASTERISK, "*"),
            (token.INT, "5"),
            (token.SEMICOLON, ";"),
            (token.INT, "5"),
            (token.LT, "<"),
            (token.INT, "10"),
            (token.GT, ">"),
            (token.INT, "5"),
            (token.SEMICOLON, ";"),
            (token.EOF, ""),
        ]

        l = lexer.Lexer(input)

        for expected_type, expected_literal in tests:
            tok = l.next_token()
            self.assertEqual(tok.type, expected_type)
            self.assertEqual(tok.literal, expected_literal)
